### PR TITLE
Release underlying resources when JS instance is GC'ed on Android try 2

### PR DESCRIPTION
**Pull request by janicduplessis**
## Summary

Reland https://github.com/facebook/react-native/pull/24767

The commit had to be reverted because it caused a crash when using remote debugging in chrome. This is normal since jsi is not available in that environment. The crash was caused by `jsContext.get()` being 0, then being dereferenced later in c++. We can simply skip initializing the blob collector in this case.

This also includes the fix from #25720 to fix a crash when using hermes.

## Changelog

[Android] [Fixed] - Release underlying resources when JS instance is GC'ed on Android

## Test Plan

Test using RN tester with jsc and hermes
Test remote debugging

### CODE REVIEW

**Comment by janicduplessis** [comment]
cc @makovkastar

File: RNTester/android/app/BUCK [hunk 1]
```diff
@@ -25,6 +25,7 @@ rn_android_library(
         react_native_target("java/com/facebook/react:react"),
         react_native_target("java/com/facebook/react/common:build_config"),
         react_native_target("java/com/facebook/react/modules/core:core"),
+        react_native_target("java/com/facebook/react/views/text:text"),
```
Comment (makovkastar) [inline RNTester/android/app/BUCK]: Why was this dependency introduced?

Comment (janicduplessis) [reply RNTester/android/app/BUCK]: I reverted https://github.com/facebook/react-native/commit/983ba630254e2ec4b8013b8ff269115e2d73ff27 to start this pr so it came from that. I’m not sure why it was it was added in the original commit. Should I just remove it?

**Comment by janicduplessis** [comment]
@makovkastar Any change you can have a look at this again?

**Comment by makovkastar** [comment]
Hi Janic, I've been busy lately with some Fabric work and didn't have time to look into it. I will try to check it this evening, but no promises.

**Comment by mcuelenaere** [comment]
Will this be backported to 0.60? This was a huge issue on our app and has caused a lot of lost engineering resources for us..

This merges with relatively little merge conflicts on 0.60-stable, see https://github.com/mcuelenaere/react-native/tree/backport/android-blob-gc-0.60

**Comment by talha-ahmed** [comment]
Any idea how soon can we get this? We have been waiting for this one for a long time. Will be great if someone can look at the broken tests so this can go in?

**Comment by makovkastar** [comment]
This PR was accepted and now there are some internal tests being run before merging this. I will try to get this merged today.

**Comment by talha-ahmed** [comment]
Thanks @makovkastar and @janicduplessis for resolving this. Looking forward to having this in the upcoming release. Any idea when that might be?

**Comment by janicduplessis** [comment]
@talha-ahmed I'll try to get it cherry-picked in the next 0.60.x release

**Comment by mcuelenaere** [comment]
To people who are interested in this, we are keeping track of these changes at https://github.com/getdelta/react-native/tree/0.60-patched (prebuilt Android binaries are available at https://github.com/getdelta/react-native/tree/0.60-patched-prebuilt)

**Comment by talha-ahmed** [comment]
This didn't get picked in 0.60.6. Can this please be released soon?

**Pull request closed without merging**